### PR TITLE
Add warning about GPU selection in Jupyter notebooks

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Due to the bug reported in #6990, warn users with `cupy` available who are working in a Jupyter notebook that they need to run `require_gpu` or `prefer_gpu` in the same cell as `spacy.load` for the model to be loaded on the correct device.

`thinc` uses `contextvars` to store the GPU selection and there is a bug in IPython (https://github.com/ipython/ipython/issues/11565) that does not preserve `contextvars` settings across cells.

### Types of change

Enhancement?

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: website/docs/usage/v3.md [hunk 1]
```diff
@@ -1168,3 +1168,15 @@ This means that spaCy knows how to initialize `my_component`, even if your
...
+
+In Jupyter notebooks, run [`prefer_gpu`](/api/top-level#spacy.prefer_gpu),
+[`require_gpu`](/api/top-level#spacy.require_gpu) or
+[`require_cpu`](/api/top-level#spacy.require_cpu) in the same cell as
+[`spacy.load`](/api/top-level#spacy.load) to ensure that the model is loaded on the correct device.
+
+Due to a bug related to `contextvars` (see the [bug
+report](https://github.com/ipython/ipython/issues/11565)), the GPU settings may
+not be preserved correctly across cells, resulting in models being loaded on
+on the wrong device or only partially on GPU.
```
Comment (svlandeg) [inline website/docs/usage/v3.md:1182]:
```suggestion
not be preserved correctly across cells, resulting in models being loaded on
the wrong device or only partially on GPU.
```

File: spacy/language.py [hunk 2]
```diff
@@ -1615,6 +1615,11 @@ def from_config(
                 or lang_cls is not cls
             ):
                 raise ValueError(Errors.E943.format(value=type(lang_cls)))
+        # Warn about require_gpu if a jupyter notebook + cupy is detected
+        if is_in_jupyter():
+            from thinc.backends.cupy_ops import CupyOps
+            if CupyOps.xp is not None:
+                warnings.warn(Warnings.W111)
```
Comment (honnibal) [inline spacy/language.py:1622]: I don't really love the idea of just displaying this warning always, no matter what the user does. We should try to only warn if the user gets it wrong.

How about this:

* We set a global variable reflecting that the GPU has been enabled in some context. This global isn't going to work with threads or contexts, it's not the "source of truth". But it does tell us that something was done.
* When we load, we check whether the following is true: a) we're in a Jupyter notebook, b) The "use gpu" global is set, c) The "use gpu" context var isn't set. If all of those are true we should know we have a problem and can warn, right?

Comment (adrianeboyd) [reply spacy/language.py:1622]: Should we move all this logic into thinc, or do you just want to handle common spacy misusage here?

**Comment by adrianeboyd** [comment]
This will fail without the thinc updates from https://github.com/explosion/thinc/pull/469.

File: spacy/language.py (hunk 2 above)
Comment (honnibal) [reply spacy/language.py]: We should move a lot of the logic into Thinc, but I don't see how to detect the different cells between `require_gpu` and `spacy.load` only in Thinc.

Comment (adrianeboyd) [reply spacy/language.py]: See: https://github.com/explosion/thinc/pull/469

Comment (adrianeboyd) [reply spacy/language.py]: No, I don't think we can detect what you've written, just that the basic thread stuff doesn't match the basic contextvars stuff. Obviously there are environments where this check wouldn't work, but it seems okay for the `is_in_jupyter` + `cupy` context?

File: website/docs/api/top-level.md [hunk 3]
```diff
@@ -158,6 +165,13 @@ if no GPU is available. If data has already been allocated on CPU, it will not
 be moved. Ideally, this function should be called right after importing spaCy
 and _before_ loading any pipelines.
 
+<Infobox variant="warning" title="Jupyter notebook usage">
```
Comment (ines) [inline website/docs/api/top-level.md:168]: Maybe we should also link to the `ipython` issue here as well (e.g. as "see this IPython issue for background on this")? If a user comes across this later and the issue happens to already be resolved, it'll be more clear why the warning is there and that it's not a general limitation of spaCy.

File: spacy/language.py [hunk 4]
```diff
@@ -1615,6 +1615,15 @@ def from_config(
                 or lang_cls is not cls
             ):
                 raise ValueError(Errors.E943.format(value=type(lang_cls)))
+        # Warn about require_gpu if a jupyter notebook + cupy + mismatched
+        # contextvars vs. thread ops are detected
+        if is_in_jupyter():
```
Comment (ines) [inline spacy/language.py:1620]: Not your fault but this kis kinda ugly so maybe we should move it into a helper function? 🤔

File: website/docs/api/top-level.md (hunk 3 above)
Comment (adrianeboyd) [reply website/docs/api/top-level.md:168]: Maybe link to the new v3 usage section that does have the link?

Comment (ines) [reply website/docs/api/top-level.md:168]: Yeah, or that 👍

File: spacy/language.py (hunk 4 above)
Comment (ines) [reply spacy/language.py:1620]: Yeah I think the helper function is worth it, especially considering the `Language.from_config` is already not the easiest method to read/follow. 

It'll hopefully all be temporary anyways.

Comment (adrianeboyd) [reply spacy/language.py:1620]: The bug report is old enough that I wouldn't have much short-term hope about an actual solution.

**Review by ines — APPROVED** [review]

**Pull request merged**